File: python/hackerrank/string/common_child.py
def common_child_slow(s1: str, s2: str) -> int:
    """
    https://www.hackerrank.com/challenges/common-child/
    """
    if len(s1) == 0 or len(s2) == 0:
        return 0
    elif s1[len(s1) - 1] == s2[len(s2) - 1]:
        return 1 + common_child_slow(s1[:len(s1)-1], s2[:len(s2)-1])
    else:
        return max(common_child_slow(s1[:len(s1)-1], s2), common_child_slow(s1, s2[:len(s2)-1]))


def common_child_memoization(s1: str, s2: str) -> int:
    """
    https://www.hackerrank.com/challenges/common-child/

    Bottom up Approach
    """
    dp = [[0 for _ in range(len(s2) + 1)] for _ in range(len(s1) + 1)]

    for i in range(1, len(s1) + 1):
        for j in range(1, len(s2) + 1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = dp[i-1][j-1] + 1
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])

    return dp[len(s1)][len(s2)]


if __name__ == "__main__":
    print("Common Child Result", common_child_memoization("AGGTAB", "GXTXAYB"))
    print("Common Child Result", common_child_memoization(
        "ABCD", "ABDC"))
    print("Common Child Result", common_child_memoization(
        "WEWOUCUIDGCGTRMEZEPXZFEJWISRSBBSYXAYDFEJJDLEBVHHKS", "FDAGCXGKCTKWNECHMRXZWMLRYUCOCZHJRRJBOAJOQJZZVUYXIC"))

File: go/hackerrank/string/common_child.go
package string

func findMax(a, b int32) int32 {
	if a > b {
		return a
	}

	return b
}

// https://www.hackerrank.com/challenges/common-child/problem
func commonChild(s1, s2 string) int32 {
	// We are going to use bottom up approach here of Longest Common Subsequence

	// Create our memoization dynamic programming
	n := len(s1) + 1
	memo := make([][]int32, n)
	for i := range memo {
		memo[i] = make([]int32, n)
	}

	for i := 1; i < n; i++ {
		for j := 1; j < n; j++ {
			if s1[i-1] == s2[j-1] {
				memo[i][j] = 1 + memo[i-1][j-1]
			} else {
				memo[i][j] = findMax(memo[i-1][j], memo[i][j-1])
			}
		}
	}

	return memo[n-1][n-1]
}
